test/integration/integration_test.go

```diff
@@ -7,54 +7,1073 @@
+	)
+
+	var (
+		emailUserNameClaim = "email"
+		ctx                = context.Background()
 	)
 
-	Context("Authenticating user to the kube api-server via trusted identity provider", func() {
-		DescribeTable("user authentication",
-			func(userToken string, expectedStatusCode int) {
+	Context("Authenticating user to the kube api-server via token from a trusted identity provider", func() {
+		createRoleBindingForUser := func(ctx context.Context, namespace, roleName, username string) *rbacv1.RoleBinding {
+			roleBinding, err := clientset.RbacV1().RoleBindings(namespace).Create(ctx, &rbacv1.RoleBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      rand.String(5),
+					Namespace: namespace,
+				},
+				Subjects: []rbacv1.Subject{{
+					Kind:     "User",
+					Name:     username,
+					APIGroup: "rbac.authorization.k8s.io",
+				}},
+				RoleRef: rbacv1.RoleRef{
+					APIGroup: "rbac.authorization.k8s.io",
+					Kind:     "Role",
+					Name:     roleName,
+				},
+			}, metav1.CreateOptions{})
+
+			Expect(err).NotTo(HaveOccurred())
+			return roleBinding
+		}
+
+		createRoleBindingForGroup := func(ctx context.Context, namespace, roleName, groupName string) *rbacv1.RoleBinding {
+			roleBinding, err := clientset.RbacV1().RoleBindings(namespace).Create(ctx, &rbacv1.RoleBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      rand.String(5),
+					Namespace: namespace,
+				},
+				Subjects: []rbacv1.Subject{{
+					Kind:     "Group",
+					Name:     groupName,
+					APIGroup: "rbac.authorization.k8s.io",
+				}},
+				RoleRef: rbacv1.RoleRef{
+					APIGroup: "rbac.authorization.k8s.io",
+					Kind:     "Role",
+					Name:     roleName,
+				},
+			}, metav1.CreateOptions{})
+
+			Expect(err).NotTo(HaveOccurred())
+			return roleBinding
+		}
+
+		deleteRoleBinding := func(ctx context.Context, roleBinding *rbacv1.RoleBinding) {
+			err := clientset.RbacV1().RoleBindings(roleBinding.Namespace).Delete(ctx, roleBinding.Name, metav1.DeleteOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		}
+
+		makeAPIServerRequestAndExpectCode := func(userToken string, expectedStatusCode int) {
+			tr := &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			}
+			client := &http.Client{Transport: tr}
+			req, err := http.NewRequest("GET", fmt.Sprintf("https://localhost:%v/api/v1/namespaces/default/pods", apiServerSecurePort), nil)
+			Expect(err).NotTo(HaveOccurred())
+			req.Header.Set("Authorization", fmt.Sprintf("Bearer %v", userToken))
+
+			Eventually(func() int {
+				res, err := client.Do(req)
+				Expect(err).NotTo(HaveOccurred())
+				return res.StatusCode
+			}, timeout, interval).Should(Equal(expectedStatusCode))
+
+		}
+
+		It("Should authenticate but not authorize user with a single registered identity provider", func() {
+			idp, err := mockidp.NewIdentityServer("mytestserver", 2)
+			Expect(err).NotTo(HaveOccurred())
+			err = idp.Start()
+			Expect(err).NotTo(HaveOccurred())
```

Comment:
You need something like:
```suggestion
			Expect(err).NotTo(HaveOccurred())
			defer idp.Stop(ctx)
```
 to ensure that the server will be stopped even when the test fail.